### PR TITLE
WIP : Provide support for Inventory Task Generation

### PR DESCRIPTION
**Pull request by rehammuzzamil**
https://github.com/OpenSRP/opensrp-plan-evaluator/issues/83

### CODE REVIEW

**Review by githengi — CHANGES_REQUESTED** [review]
Why are domain classes deleted?

File: src/main/java/org/opensrp/domain/postgres/PgStockAndProductDetails.java [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package org.opensrp.domain;
+package org.opensrp.domain.postgres;
```
Comment (githengi) [inline src/main/java/org/opensrp/domain/postgres/PgStockAndProductDetails.java:1]: Why is this moved. This class shoould not be updated so that it continues to work for implementations that don't need to access the product catalogou

File: src/main/java/org/opensrp/repository/postgres/mapper/xml/ProductCatalogueMapper.xml [hunk 2]
```diff
@@ -9,7 +9,7 @@
     -->
     <id column="unique_id" jdbcType="BIGINT" property="uniqueId" />
     <result column="product_name" jdbcType="VARCHAR" property="productName" />
-    <result column="json" jdbcType="OTHER" property="json" typeHandler="org.opensrp.repository.postgres.handler.ProductCatalogueTypeHandler" />
+    <result column="product_catalogue_json" jdbcType="OTHER" property="json" typeHandler="org.opensrp.repository.postgres.handler.ProductCatalogueTypeHandler" />
```
Comment (githengi) [inline src/main/java/org/opensrp/repository/postgres/mapper/xml/ProductCatalogueMapper.xml:12]: Can you just create a custom result for query that returns both stock and catalog

File: src/main/java/org/opensrp/repository/postgres/mapper/xml/ProductCatalogueMapper.xml [hunk 3]
```diff
@@ -124,7 +124,7 @@
       This element is automatically generated by MyBatis Generator, do not modify.
       This element was generated on Thu Oct 01 12:40:29 PKT 2020.
     -->
-    unique_id, product_name, json, server_version
+    unique_id, product_name, json AS product_catalogue_json, server_version
```
Comment (githengi) [inline src/main/java/org/opensrp/repository/postgres/mapper/xml/ProductCatalogueMapper.xml:127]: Revert this and just create a custom result for query that returns both stock and catalog

**Review by githengi — CHANGES_REQUESTED** [review]

File: src/main/java/org/opensrp/repository/postgres/mapper/custom/xml/CustomStockMetadataMapper.xml [hunk 4]
```diff
@@ -140,4 +153,23 @@
         	where stock_id = #{stockId,jdbcType=BIGINT}
     </select>
 
+	<select id="selectManyStockAndProductDetailsByServicePointId" resultMap="StockAndProductDetails">
```
Comment (rehammuzzamil) [inline src/main/java/org/opensrp/repository/postgres/mapper/custom/xml/CustomStockMetadataMapper.xml:156]: Here is the custom query responsible to fetch both stock and product catalog details @githengi

File: src/main/java/org/opensrp/repository/postgres/mapper/xml/ProductCatalogueMapper.xml (hunk 2 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/repository/postgres/mapper/xml/ProductCatalogueMapper.xml:12]: Yes, a custom query was made which returns both stock and product catalogue details

File: src/main/java/org/opensrp/repository/postgres/mapper/xml/ProductCatalogueMapper.xml (hunk 3 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/repository/postgres/mapper/xml/ProductCatalogueMapper.xml:127]: This alias was created to avoid the conflict of the same column name "json"  in both stock and product_catalogue table.

File: src/main/java/org/opensrp/repository/postgres/mapper/xml/ProductCatalogueMapper.xml (hunk 2 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/repository/postgres/mapper/xml/ProductCatalogueMapper.xml:12]: Yes, a custom query was made which returns both stock and product catalog details

File: src/main/java/org/opensrp/repository/postgres/mapper/custom/xml/CustomStockMetadataMapper.xml [hunk 5]
```diff
@@ -2,11 +2,24 @@
 <!DOCTYPE mapper PUBLIC "-//mybatis.org//DTD Mapper 3.0//EN" "http://mybatis.org/dtd/mybatis-3-mapper.dtd">
 <mapper
 	namespace="org.opensrp.repository.postgres.mapper.custom.CustomStockMetadataMapper">
+	<resultMap id="StockAndProductDetails" type="org.opensrp.domain.postgres.PgStockAndProductDetails">
```
Comment (rehammuzzamil) [inline src/main/java/org/opensrp/repository/postgres/mapper/custom/xml/CustomStockMetadataMapper.xml:5]: Yes, defined a new resultmap for the custom query.

File: src/main/java/org/opensrp/service/EventService.java [hunk 6]
```diff
@@ -299,6 +299,13 @@ public void updateEvent(Event updatedEvent) {
 		updatedEvent.setDateEdited(DateTime.now());
 		updatedEvent.setServerVersion(allEvents.getNextServerVersion());
 		allEvents.update(updatedEvent);
+		String planIdentifier = updatedEvent.getDetails() != null ? updatedEvent.getDetails().get("planIdentifier") : null;
+		if (isPlanEvaluationEnabled && planIdentifier != null) {
+			PlanDefinition plan = planRepository.get(planIdentifier);
+			if (plan.getStatus().equals(PlanDefinition.PlanStatus.ACTIVE) && (plan.getEffectivePeriod().getEnd() == null
+					|| plan.getEffectivePeriod().getEnd().isAfter(LocalDate.now().toDateTimeAtStartOfDay())))
+				taskGenerator.processPlanEvaluation(plan, username, updatedEvent);
+		}
```
Comment (githengi) [inline src/main/java/org/opensrp/service/EventService.java:308]: Move this to a method so that it not repeated on addEvent. This also needs to be invoked on `addOrUpdateEvent()`

File: src/main/java/org/opensrp/domain/postgres/PgStockAndProductDetails.java [hunk 7]
```diff
@@ -7,13 +7,10 @@
...
-@NoArgsConstructor
 @AllArgsConstructor
-public class StockObjectMetadata {
-
-	private Long date_created;
-	private String to_from;
-	private Long date_updated;
-	private long version;
+@NoArgsConstructor
+public class PgStockAndProductDetails {
```
Comment (githengi) [inline src/main/java/org/opensrp/domain/postgres/PgStockAndProductDetails.java:12]: This class is duplicate of `org.opensrp.domain.StockAndProductDetails` on path evaluator and should be deleted

**Review by githengi — CHANGES_REQUESTED** [review]

File: src/main/java/org/opensrp/domain/postgres/PgStockAndProductDetails.java (hunk 7 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/domain/postgres/PgStockAndProductDetails.java:12]: No, the entities of Stock and Product Catalogue in this domain is of Postgres while on Plan Evaluator stock and product catalogue are domain entities. @githengi 
Since the query would result in the Postgres entity, we cannot utilize the PlanEvaluator's class here.

File: src/main/java/org/opensrp/service/EventService.java (hunk 6 above)
Comment (rehammuzzamil) [reply src/main/java/org/opensrp/service/EventService.java]: Sure.

File: src/main/java/org/opensrp/repository/postgres/TaskRepositoryImpl.java [hunk 8]
```diff
@@ -390,6 +391,10 @@ public Task updateTask(Task task) {
 		return get(task.getIdentifier());
 	}
 
+	@Override
```
Comment (rehammuzzamil) [inline src/main/java/org/opensrp/repository/postgres/TaskRepositoryImpl.java:394]: @githengi  This method signature was defined by Kigamba on Plan Evaluator, therefore to avoid compilation issues. I have thrown an exception on Not Implemented. Does this look okay?

Comment (githengi) [reply src/main/java/org/opensrp/repository/postgres/TaskRepositoryImpl.java:394]: you can just implement it

Comment (githengi) [reply src/main/java/org/opensrp/repository/postgres/TaskRepositoryImpl.java:394]: leave it as it is the method should be amended to add plan filter

File: src/main/java/org/opensrp/domain/postgres/PgStockAndProductDetails.java (hunk 7 above)
Comment (githengi) [reply src/main/java/org/opensrp/domain/postgres/PgStockAndProductDetails.java:12]: the classes have the same signature and since the class is not auto-generated just use the class on PlanEvaluator's on the mapper and drop this

**Review by githengi — APPROVED** [review]

**Review by githengi — APPROVED** [review]
